docs/tutorial-advanced/outputs-and-artifacts.md

```diff
@@ -0,0 +1,640 @@
+- Less reliable than output files
+
+**When to use:** In CI/CD pipelines where you need to pass simple string values to subsequent steps. For reliability, combine with output files.
+
+### 3. Sharing Data Between Tasks (Within Same Build)
+
+While you cannot return data from `Invoke-psake` to the caller using variables, you **can** share data between tasks within the same build using script-scoped variables.
+
+```powershell title="psakefile.ps1"
+# Script-scoped variable accessible to all tasks
+$script:BuildMetadata = @{}
+
+Properties {
+    $Configuration = "Release"
+}
+
+Task Init {
+    # Store data for use by other tasks
+    $script:BuildMetadata.StartTime = Get-Date
+    $script:BuildMetadata.Configuration = $Configuration
+}
+
+Task Build -Depends Init {
+    # Access data from previous task
+    Write-Host "Started at: $($script:BuildMetadata.StartTime)"
+    Write-Host "Configuration: $($script:BuildMetadata.Configuration)"
+
+    exec { dotnet build -c $Configuration }
+
+    # Add more data
+    $script:BuildMetadata.BuildCompleted = Get-Date
+}
+
+Task Package -Depends Build {
+    # Access accumulated metadata
+    $duration = $script:BuildMetadata.BuildCompleted - $script:BuildMetadata.StartTime
+    Write-Host "Build took: $($duration.TotalSeconds) seconds"
+
+    # Package artifacts
+    exec { dotnet pack }
+}
+```
+
+**Important:** These variables are **only** accessible within the same `Invoke-psake` call. They **cannot** be accessed by the calling script.
+
+**When to use:** For passing data between tasks within the same build execution.
+
+### 4. BuildTearDown for Centralized Reporting
+
+Use `BuildTearDown` to generate summary reports or outputs that run after all tasks complete (even on failure).
+
+```powershell title="psakefile.ps1"
+# Track build metadata
+$script:BuildMetrics = @{
+    TasksExecuted = @()
+    StartTime = $null
+    EndTime = $null
+    ArtifactUrl = $null
+}
+
+BuildSetup {
+    $script:BuildMetrics.StartTime = Get-Date
+}
+
+TaskSetup {
+    $taskName = $psake.context.Peek().currentTaskName
+    $script:BuildMetrics.TasksExecuted += $taskName
+}
+
+BuildTearDown {
+    $script:BuildMetrics.EndTime = Get-Date
+    $duration = $script:BuildMetrics.EndTime - $script:BuildMetrics.StartTime
+
+    # Create summary report (IMPORTANT: Write to file, not just variables)
+    $summary = @{
+        Success = $psake.build_success
+        Duration = $duration.TotalSeconds
+        TasksExecuted = $script:BuildMetrics.TasksExecuted
+        ArtifactUrl = $script:BuildMetrics.ArtifactUrl
+        Timestamp = (Get-Date).ToString("o")
+    }
+
+    # Write to file so it's accessible after Invoke-psake completes
+    $summary | ConvertTo-Json | Set-Content ./build-summary.json
+
+    Write-Host "`n========== Build Summary ==========" -ForegroundColor Cyan
+    Write-Host "Status: $(if ($psake.build_success) { 'SUCCESS' } else { 'FAILED' })"
+    Write-Host "Duration: $($duration.TotalSeconds) seconds"
```

Comment:
`BuildTearDown`,  `BuildSetup`, are not valid commands. `TaskSetup` and `TaskTearDown` are commands, but they run before and after each task.